File: 102_Binary_Tree_Level_Order_Traversal.js
// Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).
//
// For example:
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7
// return its level order traversal as:
// [
//   [3],
//   [9,20],
//   [15,7]
// ]

var levelOrder = function(root) {
  const r = [];
  const levelOrder = (node, level=0) => {
    if (node) {
      r[level] = r[level] || [];
      r[level].push(node.val);
      levelOrder(node.left, level + 1);
      levelOrder(node.right, level + 1);
    }
  }
  levelOrder(root);
  return r;
};

使用level变量, 控制层序遍历的层数.
使用递归, 并带入level变量, 存储层序遍历的值.

File: BB/102. Binary Tree Level Order Traversal.java
Given a binary tree, return the level order traversal of its nodes values. (ie, from left to right, level by level).

For example:
Given binary tree [3,9,20,null,null,15,7],
    3
   / \
  9  20
    /  \
   15   7
return its level order traversal as:
[
  [3],
  [9,20],
  [15,7]
]


class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
      List<List<Integer>> ans = new ArrayList<>();
      if (root == null) {
        return ans;
      }
      Queue<TreeNode> queue = new LinkedList<>();
      queue.add(root);

      while (!queue.isEmpty()) {
        int size = queue.size();
        List<Integer> out = new ArrayList<>();
        for (int i = 0; i < size; i++) {
          TreeNode node = queue.poll();
          out.add(node.val);
          if(node.left != null) queue.add(node.left);
          if(node.right != null) queue.add(node.right);
        }
        ans.add(out);
      }
      return ans;

    }
}
